File: thor/src/jvmMain/kotlin/io/github/remmerw/thor/cobra/html/renderer/RAbstractCell.java
package io.github.remmerw.thor.cobra.html.renderer;

import org.eclipse.jdt.annotation.NonNull;

import java.awt.Dimension;

import io.github.remmerw.thor.cobra.html.HtmlRendererContext;
import io.github.remmerw.thor.cobra.html.domimpl.NodeImpl;
import io.github.remmerw.thor.cobra.html.style.RenderState;
import io.github.remmerw.thor.cobra.ua.UserAgentContext;

public abstract class RAbstractCell extends RBlock {

    private VirtualCell topLeftVirtualCell;

    public RAbstractCell(NodeImpl modelNode, int listNesting, UserAgentContext pcontext, HtmlRendererContext rcontext,
                         FrameContext frameContext, RenderableContainer parentContainer) {
        super(modelNode, listNesting, pcontext, rcontext, frameContext, parentContainer);
    }

    public abstract void setCellBounds(final TableMatrix.ColSizeInfo[] colSizes, final TableMatrix.RowSizeInfo[] rowSizes, final int hasBorder, final int cellSpacingX, final int cellSpacingY);

    public abstract String getWidthText();

    public abstract String getHeightText();

    public abstract int getRowSpan();

    public abstract void setRowSpan(final int rowSpan);

    public abstract int getColSpan();

    protected abstract Dimension doCellLayout(final int width, final int height, final boolean expandWidth, final boolean expandHeight,
                                              final boolean sizeOnly);

    abstract @NonNull RenderState getRenderState();

    public VirtualCell getTopLeftVirtualCell() {
        return this.topLeftVirtualCell;
    }

    public void setTopLeftVirtualCell(final VirtualCell vc) {
        this.topLeftVirtualCell = vc;
    }

    /**
     * @return Returns the virtualColumn.
     */
    public int getVirtualColumn() {
        final VirtualCell vc = this.topLeftVirtualCell;
        return vc == null ? 0 : vc.getColumn();
    }

    /**
     * @return Returns the virtualRow.
     */
    public int getVirtualRow() {
        final VirtualCell vc = this.topLeftVirtualCell;
        return vc == null ? 0 : vc.getRow();
    }

}

File: thor/src/jvmMain/kotlin/io/github/remmerw/thor/cobra/html/renderer/RAbstractCell.kt
package io.github.remmerw.thor.cobra.html.renderer

import io.github.remmerw.thor.cobra.html.HtmlRendererContext
import io.github.remmerw.thor.cobra.html.dom.NodeImpl
import io.github.remmerw.thor.cobra.html.renderer.TableMatrix.ColSizeInfo
import io.github.remmerw.thor.cobra.html.renderer.TableMatrix.RowSizeInfo
import io.github.remmerw.thor.cobra.html.style.RenderState
import io.github.remmerw.thor.cobra.ua.UserAgentContext
import java.awt.Dimension

abstract class RAbstractCell(
    modelNode: NodeImpl?,
    listNesting: Int,
    pcontext: UserAgentContext?,
    rcontext: HtmlRendererContext?,
    frameContext: FrameContext?,
    parentContainer: RenderableContainer?
) : RBlock(modelNode, listNesting, pcontext, rcontext, frameContext, parentContainer) {
    var topLeftVirtualCell: VirtualCell? = null

    abstract fun setCellBounds(
        colSizes: List<ColSizeInfo>,
        rowSizes: List<RowSizeInfo>,
        hasBorder: Int,
        cellSpacingX: Int,
        cellSpacingY: Int
    )

    abstract fun getWidthText(): String?

    abstract fun getHeightText(): String?

    abstract fun getRowSpan(): Int

    abstract fun getColSpan(): Int

    abstract fun doCellLayout(
        width: Int, height: Int, expandWidth: Boolean, expandHeight: Boolean,
        sizeOnly: Boolean
    ): Dimension?

    abstract fun getRenderState(): RenderState

    fun virtualColumn(): Int
        {
            val vc = this.topLeftVirtualCell
            return if (vc == null) 0 else vc.column
        }

    fun virtualRow(): Int {
            val vc = this.topLeftVirtualCell
            return if (vc == null) 0 else vc.row
        }

    abstract fun setColSpan(cs: Int)
    abstract fun setRowSpan(cs: Int)

}